### PR TITLE
#19978 SpecViolation should result in orderly shutdown of GraphInterp…

### PR DESCRIPTION
**Pull request by drewhk**
…reter

### CODE REVIEW

**Comment by drewhk** [comment]
Fixes #19978

File: akka-stream-tests/src/test/scala/akka/stream/impl/fusing/ActorGraphInterpreterSpec.scala [hunk 1]
```diff
@@ -331,5 +335,64 @@ class ActorGraphInterpreterSpec extends AkkaSpec {
...
+        Source.fromPublisher(filthyPublisher),
+        Source.fromPublisher(upstream))(count ⇒ Merge(count))
+        .runWith(Sink.fromSubscriber(downstream))
+
+      upstream.ensureSubscription()
+      upstream.expectCancellation()
+
+      downstream.ensureSubscription()
+      downstream.expectError() match {
+        case ise: IllegalStateException ⇒
```
Comment (2beaucoup) [inline akka-stream-tests/src/test/scala/akka/stream/impl/fusing/ActorGraphInterpreterSpec.scala]: what do you think of
```scala
val ise = the[IllegalStateException]` thrownBy downstream.expectError()
ise.getCause should (be (a [SpecViolation])
ise.getCause.getCause should (be (a [TE]) and have message "violating your spec")
```
?

Comment (drewhk) [inline akka-stream-tests/src/test/scala/akka/stream/impl/fusing/ActorGraphInterpreterSpec.scala]: nice! I have minimal ScalaTest fu

Comment (2beaucoup) [inline akka-stream-tests/src/test/scala/akka/stream/impl/fusing/ActorGraphInterpreterSpec.scala]: hehe, me too. just remembered `thrownBy` and added a little google fu. ;)

Comment (drewhk) [inline akka-stream-tests/src/test/scala/akka/stream/impl/fusing/ActorGraphInterpreterSpec.scala]: That is not ok, thrownBy only works if something is thrown

Comment (2beaucoup) [inline akka-stream-tests/src/test/scala/akka/stream/impl/fusing/ActorGraphInterpreterSpec.scala]: Ups! match != catch... Sorry then.

Comment (drewhk) [inline akka-stream-tests/src/test/scala/akka/stream/impl/fusing/ActorGraphInterpreterSpec.scala]: never mind, I refactored it a bit. Also, for me `and have` did not work, neither `should be`, but `shouldBe` did. So I deviated somewhat from your suggestion in the end.

Comment (2beaucoup) [inline akka-stream-tests/src/test/scala/akka/stream/impl/fusing/ActorGraphInterpreterSpec.scala]: :+1: nice!

**Comment by ktoso** [comment]
LGTM

File: akka-stream/src/main/scala/akka/stream/impl/fusing/ActorGraphInterpreter.scala [hunk 2]
```diff
@@ -449,7 +452,7 @@ private[stream] final class GraphInterpreterShell(
   }
 
   private var _isTerminated = false
-  def isTerminated: Boolean = _isTerminated
+  def isTerminated: Boolean = _isTerminated && canShutDown
```
Comment (viktorklang) [inline akka-stream/src/main/scala/akka/stream/impl/fusing/ActorGraphInterpreter.scala]: Under which circumstances would _isTerminated be true and canShutDown not?

File: akka-stream/src/main/scala/akka/stream/impl/fusing/ActorGraphInterpreter.scala [hunk 3]
```diff
@@ -505,7 +514,7 @@ private[stream] final class GraphInterpreterShell(
       _isTerminated = true
```
Comment (viktorklang) [inline akka-stream/src/main/scala/akka/stream/impl/fusing/ActorGraphInterpreter.scala]: On line 503, why is it safe to catch a NonFatal and just ignore it?

File: akka-stream/src/main/scala/akka/stream/impl/fusing/ActorGraphInterpreter.scala (hunk 2 above)
Comment (drewhk) [inline akka-stream/src/main/scala/akka/stream/impl/fusing/ActorGraphInterpreter.scala]: when we are still waiting on exposedPublishers and arriving subscriptions

File: akka-stream/src/main/scala/akka/stream/impl/fusing/ActorGraphInterpreter.scala (hunk 3 above)
Comment (drewhk) [inline akka-stream/src/main/scala/akka/stream/impl/fusing/ActorGraphInterpreter.scala]: yes, we are already aborting and re-aborting if abort failed makes no sense, we just give up.

Comment (drewhk) [inline akka-stream/src/main/scala/akka/stream/impl/fusing/ActorGraphInterpreter.scala]: see comment: https://github.com/akka/akka/pull/20295/files/6f2ba6ae39abbe05db4200f31ec48887486fcff2#diff-b14bd6455698812fc9b3ee7e4a5b139aL493

File: akka-stream/src/main/scala/akka/stream/impl/fusing/ActorGraphInterpreter.scala (hunk 2 above)
Comment (viktorklang) [inline akka-stream/src/main/scala/akka/stream/impl/fusing/ActorGraphInterpreter.scala]: Doesn't that mean that _isTerminated is a misnomer?

File: akka-stream/src/main/scala/akka/stream/impl/fusing/ActorGraphInterpreter.scala (hunk 3 above)
Comment (viktorklang) [inline akka-stream/src/main/scala/akka/stream/impl/fusing/ActorGraphInterpreter.scala]: Having the comment on the actual catch-and-ignore would be highly advisable, don't you agree?

File: akka-stream/src/main/scala/akka/stream/impl/fusing/ActorGraphInterpreter.scala (hunk 2 above)
Comment (drewhk) [inline akka-stream/src/main/scala/akka/stream/impl/fusing/ActorGraphInterpreter.scala]: I guess so, I try to figure out a new name.

File: akka-stream/src/main/scala/akka/stream/impl/fusing/ActorGraphInterpreter.scala (hunk 3 above)
Comment (drewhk) [inline akka-stream/src/main/scala/akka/stream/impl/fusing/ActorGraphInterpreter.scala]: Ok, I add a comment for clarity. I won't bite today for "don't you agree" and "highly advisable", but please omit them next time, they are unnecessarily condescending. It is completely fine to just ask me to put in an additional comment.

Comment (viktorklang) [inline akka-stream/src/main/scala/akka/stream/impl/fusing/ActorGraphInterpreter.scala]: Thanks. (I only added that part to not make it sound like a command, but will omit in the future, thanks for letting me know that they were not achieving the intended effect)

File: akka-stream/src/main/scala/akka/stream/impl/fusing/ActorGraphInterpreter.scala (hunk 2 above)
Comment (viktorklang) [inline akka-stream/src/main/scala/akka/stream/impl/fusing/ActorGraphInterpreter.scala]: How about "shouldShutDown"? I.e. isTerminated == shouldShutDown && canShutDown ?

File: akka-stream/src/main/scala/akka/stream/impl/fusing/ActorGraphInterpreter.scala (hunk 3 above)
Comment (drewhk) [inline akka-stream/src/main/scala/akka/stream/impl/fusing/ActorGraphInterpreter.scala]: No worries, I am in a good mood today :)

**Comment by johanandren** [comment]
I agree with Viktors two pointy-haired-boss-comments, LGTM except for that.

**Comment by patriknw** [comment]
LGTM, after final touch

**Comment by drewhk** [comment]
Fixed comments, waiting on validation.

**Pull request merged**